crates/bevy_remote/src/lib.rs

```diff
@@ -0,0 +1,877 @@
+//!             "components": [
+//!                 "bevy_transform::components::transform::Transform"
+//!             ]
+//!         }
+//!     }
+//! }
+//! ```
+//!
+//! And a response might look like this:
+//!
+//! ```json
+//! {
+//!     "components": {
+//!         "bevy_transform::components::transform::Transform": {
+//!             "rotation": { "x": 0.0, "y": 0.0, "z": 0.0, "w": 1.0 },
+//!             "scale": { "x": 1.0, "y": 1.0, "z": 1.0 },
+//!             "translation": { "x": 0.0, "y": 0.5, "z": 0.0 }
+//!         }
+//!     },
+//!     "entity": 4294967298,
+//!     "id": 0,
+//!     "status": "OK"
+//! }
+//! ```
+use std::{
+    any::TypeId,
+    net::SocketAddr,
+    sync::{Arc, Mutex},
+    thread,
+};
+
+use anyhow::{anyhow, Result as AnyhowResult};
+use bevy_app::prelude::*;
+use bevy_derive::{Deref, DerefMut};
+use bevy_ecs::{
+    component::ComponentId,
+    entity::Entity,
+    query::QueryBuilder,
+    reflect::{AppTypeRegistry, ReflectComponent},
+    system::{Commands, Res, Resource},
+    world::{EntityRef, EntityWorldMut, FilteredEntityRef, World},
+};
+use bevy_hierarchy::BuildWorldChildren as _;
+use bevy_reflect::{
+    serde::{ReflectSerializer, TypedReflectDeserializer},
+    Reflect, TypeRegistration, TypeRegistry,
+};
+use bevy_utils::{tracing::error, HashMap};
+use http_body_util::{BodyExt as _, Full};
+use hyper::{
+    body::{Bytes, Incoming},
+    service, Request, Response,
+};
+use hyper_util::{
+    rt::{TokioExecutor, TokioIo},
+    server::conn::auto::Builder,
+};
+use serde::{de::DeserializeSeed as _, Deserialize, Serialize};
+use serde_json::{value, Map, Value};
+use tokio::{
+    net::TcpListener,
+    sync::{
+        broadcast::{self, Receiver as BroadcastReceiver, Sender as BroadcastSender},
+        oneshot::{self, Sender as OneshotSender},
+    },
+    task,
+};
+
+/// The default port that Bevy will listen on.
+///
+/// This value was chosen randomly.
+pub const DEFAULT_PORT: u16 = 15702;
+
+const CHANNEL_SIZE: usize = 16;
+
+/// Add this plugin to your [`App`] to allow remote connections to inspect and modify entities.
+///
+/// By default, this is [`DEFAULT_PORT`]: 15702.
+pub struct RemotePlugin {
+    /// The port that Bevy will listen on.
+    pub port: u16,
+}
+
+/// A resource containing the port number that Bevy will listen on.
+#[derive(Resource, Reflect)]
+pub struct RemotePort(pub u16);
+
+/// A single request from a Bevy Remote Protocol client to the server,
+/// serialized in JSON.
+///
+/// The JSON payload is expected to look like this:
+///
+/// ```json
+/// {
+///     "request": "GET",
+///     "id": 0,
+///     "params": {
+///         "data": {
+///             "entity": 4294967298,
+///             "components": [
+///                 "bevy_transform::components::transform::Transform"
+///             ]
+///         }
+///     }
+/// }
+/// ```
+#[derive(Serialize, Deserialize, Clone)]
+#[serde(tag = "request", content = "params")]
+pub enum BrpRequest {
```

Comment:
I personally don't like the use of a single enum for all possible `BrpRequest` types. It makes sense for a small protocol, but we know that this will be extended to have substantially more commands, many of which will be outside of first-party control.

I propose instead representing a request as a verb and some unknown payload. The user could then register handlers as just systems with an input `In: DeserializedOwned + 'static` and an output `Out: Serialize + 'static`. Adding a handler would be a matter of passing the `SystemId` to the message handling resource:

```rust
impl App {
    fn add_brp_handler<In, Out>(
        &mut self,
        verb: impl Into<String>,
        system_id: SystemId<In, Out>,
    ) -> &mut Self
    where
        In: DeserializeOwned + 'static,
        Out: Serialize + 'static,
    {
        let verb: Arc<str> = Arc::from(verb.into());

        let handler = Box::new(move |world: &mut World, request: BrpRequest| {
            let input: In = request.deserialize_params().unwrap();
            let output = world.run_system_with_input(system_id, input).unwrap();
            let response = BrpResponse::serialize_response(request, output).unwrap();

            Ok(response)
        });

        let mut registry = self
            .world_mut()
            .get_resource_or_insert_with::<BrpRegistry>(Default::default);

        registry.handlers.insert(verb, handler);

        self
    }
}

// In user code

fn damage_entity(request: In<DamageParams>, query<&mut Health>) {
    if let Ok(mut health) = query.get_mut(request.entity) {
        health.0 -= request.damage;
    }
}

app.add_brp_handler("DAMAGE", damage_entity);
```

The main-loop would then just check the verb against its list of handlers, and delegate from there. Personally, I think this would be a massive ergonomics win for end users.

Reply:
Yeah, I was thinking about something like this. I was thinking maybe require that all app-specific verbs be prefixed with `X-`? That way we reserve the "main" namespace for future expansion.

Reply:
I actually think we want to replace the `verb` concept with a path anyway looking at it in more detail. This would allow Bevy to register all of its first-party commands under `bevy/*` (with folders for specific command groups perhaps?). Command resolution could then walk backwards over the provided path, allowing an end user to just send `GET` and the BRP handler would choose `bevy/GET` if it's the only `GET` handler registered.

That would allow users to use short names, and editors to use full names.

Reply:
I'd prefer to avoid ambiguity; if we're going to namespace verbs, we should just do that.

Reply:
This is now done. I didn't do any namespacing due to the debate not being settled.